File: app/src/main/java/kim/uno/ui/adapter/BaseRecyclerAdapter.java
package kim.uno.ui.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;

import kim.uno.item.RecyclerItem;
import kim.uno.ui.holder.BaseViewHolder;

public abstract class BaseRecyclerAdapter<T extends RecyclerItem> extends RecyclerView.Adapter<BaseViewHolder> {

    private Context mContext;
    private ArrayList<T> mItems;

    public BaseRecyclerAdapter(Context context) {
        mContext = context;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int type) {
        return onCreateNewHolder(parent, type);
    }

    public abstract BaseViewHolder onCreateNewHolder(ViewGroup parent, int type);

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        holder.onBindView(getItem(position) != null ? getItem(position) : null, position);
    }

    @Override
    public int getItemViewType(int position) {
        return getItem(position) != null ? getItem(position).getViewType() : 0;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public Context getContext() {
        return mContext;
    }

    @Override
    public int getItemCount() {
        return mItems != null ? mItems.size() : 0;
    }

    public T getItem(int position) {
        return getItemCount() > position ? mItems.get(position) : null;
    }

    public ArrayList<T> getItems() {
        return mItems;
    }

    public void setItems(ArrayList<T> items) {
        setItems(items, true);
    }

    public void setItems(ArrayList<T> items, boolean notify) {
        mItems = items;
        if (notify) {
            notifyDataSetChanged();
        }
    }

    public void addItems(ArrayList<T> items) {
        addItems(items, true);
    }

    public void addItems(ArrayList<T> items, boolean notify) {
        if (mItems == null) mItems = new ArrayList<>();
        ArrayList<T> mergeItems = new ArrayList<>(mItems);
        mergeItems.addAll(items);
        mItems = mergeItems;
        if (notify) {
            if (mItems.size() - items.size() == 0) {
                notifyDataSetChanged();
            } else {
                notifyItemRangeInserted(mItems.size() - items.size(), mItems.size());
            }
        }
    }

    public void addItem(T item) {
        if (mItems == null) mItems = new ArrayList<>();
        ArrayList<T> mergeItems = new ArrayList<>(mItems);
        mergeItems.add(item);
        mItems = mergeItems;
    }

    public void clear() {
        if (mItems != null) mItems.clear();
    }
}

File: base/src/main/kotlin/kim/uno/kotlin/base/ui/BaseRecyclerAdapter.kt
package kim.uno.kotlin.base.ui

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import kim.uno.kotlin.base.item.RecyclerItem
import java.util.*

abstract class BaseRecyclerAdapter<T : RecyclerItem>(val context: Context) : RecyclerView.Adapter<BaseViewHolder<T>>() {

    private var mItems: ArrayList<T>? = null

    override fun onCreateViewHolder(parent: ViewGroup, type: Int): BaseViewHolder<T> = onCreateNewHolder(parent, type)

    abstract fun onCreateNewHolder(parent: ViewGroup, type: Int): BaseViewHolder<T>

    override fun onBindViewHolder(holder: BaseViewHolder<T>, position: Int) {
        holder.onBindView(getItem(position), position)
    }

    override fun getItemViewType(position: Int): Int = getItem(position)?.getViewType()?: 0

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getItemCount(): Int = mItems?.size?: 0

    fun getItem(position: Int): T? = mItems?.getOrNull(position)

    fun setItems(items: ArrayList<T>) = setItems(items, true)

    open fun setItems(items: ArrayList<T>, notify: Boolean) {
        mItems = items
        if (notify) {
            notifyDataSetChanged()
        }
    }

    fun addItems(items: ArrayList<T>) = addItems(items, Integer.MAX_VALUE)

    fun addItems(items: ArrayList<T>, notify: Boolean) = addItems(items, Integer.MAX_VALUE, notify)

    fun addItems(items: ArrayList<T>, position: Int) = addItems(items, position, true)

    open fun addItems(items: ArrayList<T>, position: Int, notify: Boolean) {
        if (mItems == null) mItems = ArrayList<T>()

        val mergeItems = ArrayList(mItems?: ArrayList<T>())
        mergeItems.addAll(if (position < 0) 0 else if (position > itemCount) itemCount else position, items)
        mItems = mergeItems
        if (notify) {
            notifyItemRangeInserted(itemCount - items.size, itemCount)
        }
    }

    fun addItem(item: T) = addItem(item, Integer.MAX_VALUE)

    open fun addItem(item: T, position: Int) {
        if (mItems == null) mItems = ArrayList<T>()
        val mergeItems = ArrayList(mItems?: ArrayList<T>())
        mergeItems.add(if (position < 0) 0 else if (position > itemCount)  itemCount else position, item)
        mItems = mergeItems
    }

    fun clear() {
        mItems?.clear()
    }

}
